clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java

```diff
@@ -205,7 +206,7 @@ public void abortTransaction() throws ProducerFencedException {
         this.transactionInFlight = false;
     }
 
-    private void verifyProducerState() {
+    private synchronized void verifyProducerState() {
```

Comment:
why do we need to `synchronized` a private method?

I was just looking into `trunk` and it seems to be `synchronized` there, too. Also, some methods are `synchronized` others are not, but I could not see any pattern. Should we just `synchronized` all public method? Performance should not be a concern for `MockProducer`. On the other hand, when would a producer be used by multiple threads? Neither `KafkaConsumer` nor `KafkaProducer` are thread-safe by design. Why does `MockProducer` need to be thread-safe?

Reply:
I added it in the prior PRs because findbugs complained that some, but not all, accesses of the "am I fenced" field are synchronized. I didn't bother verifying that it's necessary in the older branches, but I can check it out real quick.

Reply:
I don't know why anything in here is synchronized; that's just the way I found it.

Reply:
I see. Maybe ok for older branches to leave as is. But might be wroth to cleanup in `trunk`?

Reply:
To be fair: `fenceProducer()` is not part of public producer API but added to `MockProducer` and it might be called from a different thread (ie, the main testing thread while `StreamThread` used the producer, too). Because all public methods within `MockProducer` call `verifyState()` it makes sense what findbugs complains about. Did miss this connection before. So maybe drop latest commit :) 

Sorry for the forth and back.